proxy/src/outbound.rs

```diff
@@ -54,6 +54,60 @@ impl<B> Outbound<B> {
             bind_timeout,
         }
     }
+
+    const DEFAULT_PORT: Option<u16> = Some(80);
+
+    /// TODO: Return error when `HostAndPort::normalize()` fails.
+    /// TODO: Use scheme-appropriate-default
+    fn normalize(authority: &http::uri::Authority) -> Option<HostAndPort> {
+        HostAndPort::normalize(authority, Self::DEFAULT_PORT).ok()
+    }
+
+    /// Determines the logical host:port of the request.
+    ///
+    /// 1. If the parsed URI includes an authority, use that.
+    /// 2. Otherwise, try to load the authority from the `Host` header.
+    ///
+    /// The port is either parsed from the authority or a default of 80 is used.
+    fn host_port(req: &http::Request<B>) -> Option<HostAndPort> {
+        req.uri()
+            .authority_part()
+            .and_then(Self::normalize)
+            .or_else(|| {
+                h1::authority_from_host(req)
+                    .and_then(|h| Self::normalize(&h))
+            })
+    }
+
+    /// Determines the destination to use in the router.
+    ///
+    /// 1. If the request was destined to a logic (named) host, route it by the logical
+    ///    hostname and port.
+    /// 2. Otherwise, route the request to the exact concrete destination it was
+    ///    already headed to, as determined by the SO_ORIGINAL_DST socket option.
+    ///
+    /// If the request has no logical host destination and the SO_ORIGINAL_DST socket
+    /// option has not been set or is unavailable on this platform, the request is not
+    /// recognized.
+    ///
+    /// The SO_ORIGINAL_DST socket option is typically set by iptables(8) in i.e.
+    /// containerized environments like Kubernetes (as configured by the proxy-init
+    /// program).
+    fn destination(req: &http::Request<B>) -> Option<Destination> {
+        match Self::host_port(req) {
+            Some(HostAndPort { host: Host::DnsName(dns_name), port }) => {
+                Some(Destination::Hostname(DnsNameAndPort { host: dns_name, port }))
+            }
+
+            Some(HostAndPort { host: Host::Ip(_), .. }) |
```

Comment:
I don't think this is right and I'm surprised the old code was still doing this too, since I'm sure I wrote a PR to change this behavior: If the request authority is explicitly an IP address then we should route based on that address and in particular that connection should be reusable (persistent) unlike the SO_ORIGINAL_DST case.

Reply:
Yes, that was surprising to me as well. I will address this here.